src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs

```diff
@@ -205,6 +207,63 @@ private void EnsureDone()
                 }
             }
 
+            private static bool ShouldSaveToTemp(Expression expression)
+            {
+                if (expression == null)
+                    return false;
+
+                // Some expressions have no side-effects and don't have to be
+                // stored into temporaries, e.g.
+                //
+                //     xs[0] = try { ... }
+                //           |
+                //           v
+                //        t0 = xs
+                //        t1 = 0            // <-- this is redundant
+                //        t2 = try { ... }
+                //    t0[t1] = t2
+                //           |
+                //           v
+                //        t0 = xs
+                //        t1 = try { ... }
+                //     t0[0] = t1
+
+                switch (expression.NodeType)
+                {
+                    // Emits ldnull, ldc, initobj, closure constant access, etc.
+                    case ExpressionType.Constant:
+                    case ExpressionType.Default:
+                        return false;
+
+                    // Emits calls to pure RuntimeOps methods with immutable arguments
+                    case ExpressionType.Quote:
+                    case ExpressionType.RuntimeVariables:
+                        return false;
+
+                    case ExpressionType.MemberAccess:
+                        var member = (MemberExpression)expression;
+                        var field = member.Member as FieldInfo;
+                        if (field != null)
+                        {
+                            // Emits ldc for the raw value of the field
+                            if (field.IsLiteral)
+                                return false;
+
+                            // For read-only fields we could save the receiver, but
+                            // that's more involved, so we'll just handle static fields
+                            if (field.IsInitOnly && field.IsStatic)
+                                return false;
```

Comment:
I did some more experiments to simulate the behavior of deferred execution of the type initializer and the effects of the stack spilling optimization in Roslyn and here. Consider this code:

```csharp
using System;
using System.Threading.Tasks;

class C
{
    static void Main(string[] args)
    {
        Console.WriteLine("Main");
        if (args[0] == "async")
            Async();
        else
            Sync();
    }

    static void Async() => Console.WriteLine(AsyncCore().Result);
    static void Sync()  => Console.WriteLine( SyncCore()       );

    static async Task<string> AsyncCore() => X.Foo + await Task.FromResult(Bar());
    static            string  SyncCore()  => X.Foo +                       Bar();

    static string Bar()
    {
        Console.WriteLine("Bar");
        return "bar";
    }
}

class X
{
    public static readonly string Foo = Get();

    static string Get()
    {
        Console.WriteLine("Foo");
        return "foo";
    }
}
```

The sync and async cases should exhibit the same evaluation order, but the `async` method needs to spill the stack prior to the `await`. The optimization gives `X.Foo` an exemption to spilling into a temporary local.

Results of running this are as follows:

| Flavor | CoreCLR | .NET 4.5 |
|---|---|---|
| Sync | `Main Foo Bar` | `Main Foo Bar` |
| Async | `Main Bar Foo` | `Main Foo Bar` |